psi4/src/psi4/libfock/DirectDFJK.cc

```diff
@@ -0,0 +1,1600 @@
+		}
+
+	}
+    n_col_last_ = C_left_ao_[0]->ncol();
+}
+
+void DirectDFJK::postiterations() {
+}
+
+// Determines the maximum amount of memory that DirectDFJK will be using
+//   at any one time. To be sure, the memory used to construct AO blocks
+//   will only be allocated during the compute_JK() method, so it will
+//   be available during other procedures. 
+// Uses blocking and sparsity information to set total_needs which is 
+//   returned by memory_estimate.
+void DirectDFJK::our_needs(){
+	free_memory_ = memory_;
+	size_t blocks;
+
+	size_t our_needs = 0;
+	double charges_f = 0.0;
+	size_t charges_z;
+	std::shared_ptr<Molecule> mol_ptr = primary_->molecule();
+	
+	for (int atom_iter = 0; atom_iter < mol_ptr->natom(); atom_iter++)
+		charges_f+=mol_ptr->fZ(atom_iter);
+
+	charges_f = charges_f/2.0;
+
+	charges_z = static_cast<size_t>(charges_f);
+	++charges_z;
+
+	//D
+	our_needs += nbf_*nbf_;
+	free_memory_-=nbf_*nbf_;
+
+	//C_a
+	our_needs += nbf_*charges_z;
+	free_memory_-=nbf_*charges_z;
+	//C_b
+	our_needs += nbf_*charges_z;
+	free_memory_-=nbf_*charges_z;
+	//K
+   	our_needs += nbf_*nbf_;	
+	free_memory_-=nbf_*nbf_;
+   	//J
+   	our_needs += nbf_*nbf_;
+	free_memory_-=nbf_*nbf_;
+	//[ J^{-1} ]
+	our_needs += naux_*naux_;
+	free_memory_-=naux_*naux_;
+	//[ J^{-\frac{1}{2} ]
+	our_needs += naux_*naux_;
+	free_memory_-=naux_*naux_;
+	blocks = (2* (nbf_ * nbf_ *naux_))/free_memory_ +1;
+	ABX_block_size_ = ((naux_/blocks)+1)*nbf_*nbf_;
+
+    total_needs_ = our_needs;// + 2*ABX_block_size_;
+}
+size_t DirectDFJK::memory_estimate(){
+
+	if (Qpq_) {
+		prepare_Q_blocks();
+	}
+
+	our_needs();
+	return total_needs_;
+}
+
+void DirectDFJK::prepare_metric_power(double power){
+	SharedMatrix my_met;
+	auto coul_met = std::make_shared<FittingMetric>(auxiliary_, true);
+	coul_met->form_fitting_metric();
+	my_met = coul_met->get_metric();
+	if ( !(1e-13 > fabs(1.0 - power)) ){
+		my_met->power(power, condition_);
+	}
+	
+	met_cols_.push_back(my_met->ncol());
+	met_rows_.push_back(my_met->nrow());
+	metric_.push_back(my_met);
+	met_powers_.push_back(power);
+
+    while (power < 0.0) {
+        power = power + 1.0;
+    }
+}
+
+double* DirectDFJK::get_metric_power(double power){
+	bool pow_on = false;
+	double ret_power;
+	size_t met_ind;
+	for (size_t met_iter = 0; met_iter < metric_.size(); met_iter++){
+		if (std::fabs(power - met_powers_[met_iter]) < condition_ ){
+			met_ind = met_iter;
+			pow_on = true;
+		}
+	}
+	if (!pow_on) {
+		met_ind = metric_.size();
+		SharedMatrix my_met;
+		auto coul_met = std::make_shared<FittingMetric>(auxiliary_, true);
+		coul_met->form_fitting_metric();
+		my_met = coul_met->get_metric();
+		my_met->power(power, condition_);
+		metric_.push_back(my_met);
+		met_powers_.push_back(power);
+	}
+	return metric_[met_ind]->pointer()[0];
+}
+
+// Computes the LU decomposition of the coulomb metric and stores it in
+//   member vector CMPQ_LU_.
+//   pivots are storec in vector PERMUTE_
```

Comment:
What's this used for?

Reply:
It's used for constructing the Coulomb Matrix

Reply:
Hm, not sure I follow. If you have the half-inverse overlap already, you don't need anything more.

Reply:
In line with Weigend's formulation, I'm constructing the coulomb matrix with a transform against the full inverse of the coulomb metric. I had problems with conditioning, so I went with the LU decomposition with partial pivoting to avoid having to rewrite the code more than once.

Reply:
@hokru ... and that's why porting `FittingMetric` to using `orthog.cc` would be good, since the pivoted Cholesky in `orthog.cc` has no problem with overcomplete sets.